sync/group.go

```diff
@@ -0,0 +1,95 @@
+package sync
+
+import (
+	"context"
+	"sync"
+)
+
+// A Group is a collection of goroutines working on subtasks that are part of
+// the same overall task.
+//
+// Use NewGroup to create a new group.
+type Group struct {
+	ctx     context.Context
+	cancel  context.CancelCauseFunc
+	wg      sync.WaitGroup
+	sem     chan struct{}
+	errOnce sync.Once
+	err     error
+}
+
+// NewGroup returns a new group and an associated Context derived from ctx.
+//
+// The derived Context is canceled the first time a function passed to Go
+// returns a non-nil error or the first time Wait returns, whichever occurs
+// first.
+//
+// limit < 1 means no limit on the number of active goroutines.
+func NewGroup(ctx context.Context, limit int) (*Group, context.Context) {
+	ctx, cancel := context.WithCancelCause(ctx)
+	g := &Group{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	if limit > 0 {
+		g.sem = make(chan struct{}, limit)
+	}
+	return g, ctx
+}
+
+// Go calls the given function in a new goroutine.
+// It blocks until the new goroutine can be added without the number of
+// active goroutines in the group exceeding the configured limit.
+//
+// The first call to return a non-nil error cancels the group's context.
+// The error will be returned by Wait.
+//
+// If the group was created by calling NewGroup with limit < 1, there is no
+// limit on the number of active goroutines.
+//
+// If the group's context is canceled, routines that have not executed yet due to the limit won't be executed.
+// Additionally, there is a best effort not to execute `f()` once the context is canceled
+// and that happens whether or not a limit has been specified.
+func (g *Group) Go(f func() error) {
```

Comment:
It's not necessary that the same context be passed.
In that case, we don't really know which context cancelled.

Reply:
Makes sense👍.
Just wanted to stick to `errGroup`'s interface, didn't think of it this way.

Reply:
Personally I'd keep it the same as the original interface so that we can reuse things without refactoring. The group is taking care of this for you. The only difference is not scheduling **NEW** routines. And if you use this group is only because you want that feature so it's pretty clear already, right? Wdyt?

Reply:
@lvrach if we call it `EagerGroup` and change the comment on both the `type Group struct` and the constructor to make clear that the only difference with the std errgroup is that it doesn't execute `f()` upon error/cancellation, then maybe we can keep it with the same interface as the original. Wdyt?

Reply:
I see your point, it was subjective non-blocking comment. Let's go with the drop in replacement approach.